Rubberduck.Parsing/Grammar/VBACodeStringParser.cs

```diff
@@ -0,0 +1,65 @@
+﻿using System;
+using Antlr4.Runtime;
+using Antlr4.Runtime.Atn;
+using Antlr4.Runtime.Tree;
+using NLog;
+using Rubberduck.Parsing.Symbols.ParsingExceptions;
+using Rubberduck.Parsing.VBA;
+
+namespace Rubberduck.Parsing.Grammar
+{
+    /// <summary>
+    /// The class is meant to be used for where we need to do rewriting without actually
+    /// rewriting an existing module in a VBA project. The result is completely free-floating
+    /// and it is the caller's responsibility to persist the finalized results into one of
+    /// actual code. This is useful for creating a code preview for refactoring.
+    /// </summary>
+    public class VBACodeStringParser
+    {
+        private static readonly Logger Logger = LogManager.GetCurrentClassLogger();
+        private readonly CommonTokenStream tokenStream;
+        public TokenStreamRewriter Rewriter { get => new TokenStreamRewriter(tokenStream); }
```

Comment:
I originally wanted to use `IModuleRewriter` however, this is not a compatible interface. Additionally, to use it requires having a `ICodeModule`, which I don't have. Suggestions?

Reply:
The `IModuleRewriter` interface doesn't mandate an `ICodeModule`, only the `ModuleRewriter` implementation does, because its job is specifically to rewrite the code in an actual `ICodeModule`. I'd suggest making a new implementation of the interface, that can work with arbitrary parse trees and output a simple string. Problem being that `Rewrite` doesn't output anything since it's assumed that it has the side-effects one would expect of a *rewriter*. A new `PreviewBoxRewriter` implementation could still use the `GetText` method to supply the rewritten output to its caller.

Reply:
TBH the base implementation could be refactored into a new, more general-purpose abstract class.

Reply:
IDK if I can refactor this into abstract -- what'd be not common to something that takes a parse tree from an arbitrary string? Unless we have other use cases, YAGNI principle applies, I think.

Reply:
I mean a lot of the implementation is mere boilerplate that would be copied in all implementations; that's why I derived the attribites rewriter and made the `Rewrite` method virtual: at the end of the day all rewriters do is manipulating a TSR, pretty much only the rewrite method is implementation-specific. With a virtual no-op `Rewrite` method the existing rewriters could inherit most of the implementation. That said the attributes rewriter will be going the way of the dodo soon anyway.